exchanges/huobi/huobi_websocket.go

```diff
@@ -75,427 +73,233 @@ var subscriptionNames = map[string]string{
-		if err != nil {
-			log.Errorf(log.ExchangeSys,
-				"%v - authenticated dial failed: %v\n",
-				h.Name,
-				err)
-		}
-		err = h.wsLogin(context.TODO())
-		if err != nil {
-			log.Errorf(log.ExchangeSys,
-				"%v - authentication failed: %v\n",
-				h.Name,
-				err)
-			h.Websocket.SetCanUseAuthenticatedEndpoints(false)
-		}
-	}
-
-	h.Websocket.Wg.Add(1)
-	go h.wsReadData()
-	return nil
-}
-
-func (h *HUOBI) wsDial(dialer *websocket.Dialer) error {
-	err := h.Websocket.Conn.Dial(dialer, http.Header{})
-	if err != nil {
-		return err
-	}
 	h.Websocket.Wg.Add(1)
-	go h.wsFunnelConnectionData(h.Websocket.Conn, wsMarketURL)
-	return nil
-}
+	go h.wsReadMsgs(h.Websocket.Conn)
 
-func (h *HUOBI) wsAuthenticatedDial(dialer *websocket.Dialer) error {
-	if !h.IsWebsocketAuthenticationSupported() {
-		return fmt.Errorf("%v AuthenticatedWebsocketAPISupport not enabled",
-			h.Name)
-	}
-	err := h.Websocket.AuthConn.Dial(dialer, http.Header{})
-	if err != nil {
-		return err
+	h.Websocket.SetCanUseAuthenticatedEndpoints(false)
+	if h.IsWebsocketAuthenticationSupported() {
+		if err := h.wsAuthConnect(ctx); err != nil {
+			return fmt.Errorf("error authenticating websocket: %w", err)
+		}
+		h.Websocket.Wg.Add(1)
+		go h.wsReadMsgs(h.Websocket.AuthConn)
 	}
 
-	h.Websocket.Wg.Add(1)
-	go h.wsFunnelConnectionData(h.Websocket.AuthConn, wsAccountsOrdersURL)
 	return nil
 }
 
-// wsFunnelConnectionData manages data from multiple endpoints and passes it to
-// a channel
-func (h *HUOBI) wsFunnelConnectionData(ws stream.Connection, url string) {
+// wsReadMsgs reads and processes messages from a websocket connection
+func (h *HUOBI) wsReadMsgs(s stream.Connection) {
 	defer h.Websocket.Wg.Done()
 	for {
-		resp := ws.ReadMessage()
-		if resp.Raw == nil {
+		msg := s.ReadMessage()
+		if msg.Raw == nil {
 			return
 		}
-		comms <- WsMessage{Raw: resp.Raw, URL: url}
+
+		if err := h.wsHandleData(msg.Raw); err != nil {
+			h.Websocket.DataHandler <- err
+		}
 	}
 }
-
-// wsReadData receives and passes on websocket messages for processing
-func (h *HUOBI) wsReadData() {
-	defer h.Websocket.Wg.Done()
-	for {
-		select {
-		case <-h.Websocket.ShutdownC:
-			select {
-			case resp := <-comms:
-				err := h.wsHandleData(resp.Raw)
-				if err != nil {
-					select {
-					case h.Websocket.DataHandler <- err:
-					default:
-						log.Errorf(log.WebsocketMgr,
-							"%s websocket handle data error: %v",
-							h.Name,
-							err)
-					}
-				}
-			default:
-			}
-			return
-		case resp := <-comms:
-			err := h.wsHandleData(resp.Raw)
-			if err != nil {
-				h.Websocket.DataHandler <- err
+func (h *HUOBI) wsHandleData(respRaw []byte) error {
+	for _, op := range []string{wsSubOp, wsUnsubOp} {
+		key := op + "bed" // subbed, unsubbed
+		if ch, err := jsonparser.GetString(respRaw, key); err == nil {
+			if !h.Websocket.Match.IncomingWithData(op+":"+ch, respRaw) {
+				return fmt.Errorf("%w: %s:%s", stream.ErrNoMessageListener, op, ch)
 			}
+			return nil
 		}
 	}
```

Comment:
Perhaps there's something I've missed, but I don't see the point to this loop when you have a handler for it that performs the same function below. It also doesn't force every single message on the ws into a loop - costing about 90ns per message

```
case wsSubOp, wsUnsubOp:
			return h.wsHandleV2subResp(action, respRaw)
```

Reply:
It's the V1 sub support, so the section you highlighted isn't relevant.
However it's also entirely superfluous because the id is in there and it's working, so ... remove this section and everything is still fine.

Fixed https://github.com/gbjk/gocryptotrader/commit/6938c5573